src/porepy/numerics/time_step_control.py

```diff
@@ -180,110 +291,233 @@ def __repr__(self) -> str:
-        # Note that iterations is not really used here. So, as long as the passed
-        # recompute_solution==True and recomputation_attempts < max_recomp_attempts, the user
-        # can pass any number of iterations. This, in principle, allows for more flexibility,
-        # in the sense that we are no restricting the recomputation criteria to only reaching
-        # the maximum number of iterations, even though that is the primary intended usage.
-        if recompute_solution and self._recomp_num < self.recomp_max:
-            self.time -= self.dt
-            self.dt *= self.recomp_factor
-            self._recomp_num += 1
-            if self._print_info:
-                s = "Solution did not converge and will be recomputed."
-                s += f" Recomputing attempt #{self._recomp_num}. Next dt = {self.dt}."
-                print(s)
-            if self.dt < self.dt_min:
-                self.dt = self.dt_min
-                if self._print_info:
-                    print(
-                        f"Calculated dt < dt_min. Using dt_min = {self.dt_min} instead."
-                    )
-            return self.dt
-        elif not recompute_solution:  # we reach convergence, set recomp_num to zero
-            self._recomp_num = 0
-        else:  # number of recomputing attempts has been exhausted
-            msg = f"Solution did not converge after {self.recomp_max} recomputing attempts."
-            # TODO: Should this be a RunTimeError perhaps?
-            raise ValueError(msg)
+        # If the time step is constant, always return that value
+        if self.is_constant:
+            return self.dt_init
 
-        # If iters < max_iter. Proceed to determine the next time step using the
-        # following criteria.
-        # If iters is less than the lower optimal iteration range "iter_low", we can relax
-        # the time step, and multiply by a lower multiplication factor greater than 1,
-        # i.e., "factor_low". If the number of iterations is greater than the upper optimal
-        # iteration range "iter_upp", we have to decrease the time step by multiplying by an
-        # upper multiplication factor smaller than 1, i.e., "factor_upp". If neither of these
-        # situations occur, then the number iterations lies between the optimal iteration
-        # range and the time step remains unchanged.
-        if iterations <= self.iter_low:
-            self.dt = self.dt * self.iter_low_factor
+        # Adapt time step
+        if not recompute_solution:
+            self._adaptation_based_on_iterations(iterations)
+        else:
+            self._adaptation_based_on_recomputation()
+
+        # Correct time step
+        self._correction_based_on_dt_min()
+        self._correction_based_on_dt_max()
+        self._correction_based_on_schedule()
+
+        return self.dt
+
+    def _adaptation_based_on_iterations(self, iterations) -> None:
```

Comment:
```suggestion
    def _adaptation_based_on_iterations(self, iterations: int) -> None:
```

Reply:
Indeed, mypy actually complained about this.